contracts/EnhancedStepFunctions.sol

```diff
@@ -0,0 +1,257 @@
+        if (command == 15) {
+            (stepXValue, stepYValue) = decodeStepFunctionData(tokenData[token].sellRateImbalanceStepFunction.x[param]);
+            return stepYValue;
+        }
+
+        revert();
+    }
+
+    /* solhint-disable function-max-lines */
+    function getRate(ERC20 token, uint currentBlockNumber, bool buy, uint qty) public view returns(uint) {
+        // check if trade is enabled
+        if (!tokenData[token].enabled) return 0;
+        if (tokenControlInfo[token].minimalRecordResolution == 0) return 0; // token control info not set
+
+        // get rate update block
+        bytes32 compactData = tokenRatesCompactData[tokenData[token].compactDataArrayIndex];
+
+        uint updateRateBlock = getLast4Bytes(compactData);
+        if (currentBlockNumber >= updateRateBlock + validRateDurationInBlocks) return 0; // rate is expired
+        // check imbalance
+        int totalImbalance;
+        int blockImbalance;
+        (totalImbalance, blockImbalance) = getImbalance(token, updateRateBlock, currentBlockNumber);
+
+        // calculate actual rate
+        int imbalanceQty;
+        int extraBps;
+        int8 rateUpdate;
+        uint rate;
+
+        if (buy) {
+            // start with base rate
+            rate = tokenData[token].baseBuyRate;
+
+            // add rate update
+            rateUpdate = getRateByteFromCompactData(compactData, token, true);
+            extraBps = int(rateUpdate) * 10;
+            rate = addBps(rate, extraBps);
+
+            // compute token qty
+            qty = getTokenQty(token, qty, rate);
+            imbalanceQty = int(qty);
+
+            // add imbalance overhead
+            extraBps = executeStepFunction(tokenData[token].buyRateImbalanceStepFunction, totalImbalance, totalImbalance + imbalanceQty);
+            rate = addBps(rate, extraBps);
+            totalImbalance += imbalanceQty;
+        } else {
+            // start with base rate
+            rate = tokenData[token].baseSellRate;
+
+            // add rate update
+            rateUpdate = getRateByteFromCompactData(compactData, token, false);
+            extraBps = int(rateUpdate) * 10;
+            rate = addBps(rate, extraBps);
+
+            // compute token qty
+            imbalanceQty = -1 * int(qty);
+
+            // add imbalance overhead
+            extraBps = executeStepFunction(tokenData[token].sellRateImbalanceStepFunction, totalImbalance + imbalanceQty, totalImbalance);
+            rate = addBps(rate, extraBps);
+            totalImbalance += imbalanceQty;
+        }
+
+        if (abs(totalImbalance) >= getMaxTotalImbalance(token)) return 0;
+        if (abs(blockImbalance + imbalanceQty) >= getMaxPerBlockImbalance(token)) return 0;
+
+        return rate;
+    }
+
+    // Override function getImbalance to fix https://github.com/KyberNetwork/smart-contracts/issues/240
+    function getImbalancePerToken(ERC20 token, uint rateUpdateBlock, uint currentBlock)
```

Comment:
Override function with checking overflow, so we don't touch to `VolumeImbalanceRecorder` contract

Reply:
Added test for this func, also check for overflow